File: flutter/lib/market_data/market_refresh_request_planner.dart
import "../domain/holdings_models.dart";
import "../domain/market_normalization.dart";

enum MarketRefreshTrackedSymbolKind { market, currency }

class MarketRefreshTrackedSymbolRequest {
  const MarketRefreshTrackedSymbolRequest({required this.symbol, required this.kind});

  final String symbol;
  final MarketRefreshTrackedSymbolKind kind;
}

class MarketRefreshRequestPlanner {
  static List<String> trackedSymbols({
    required List<DashboardWidget> activeWidgets,
    required List<Holding> holdings,
    required CurrencyWidgetSetting currencySettings,
    required MarketTickerSetting marketTickerSettings,
  }) {
    return trackedRequests(
      activeWidgets: activeWidgets,
      holdings: holdings,
      currencySettings: currencySettings,
      marketTickerSettings: marketTickerSettings,
    ).map((MarketRefreshTrackedSymbolRequest e) => e.symbol).toList();
  }

  static List<MarketRefreshTrackedSymbolRequest> trackedRequests({
    required List<DashboardWidget> activeWidgets,
    required List<Holding> holdings,
    required CurrencyWidgetSetting currencySettings,
    required MarketTickerSetting marketTickerSettings,
  }) {
    final List<MarketRefreshTrackedSymbolRequest> requests = <MarketRefreshTrackedSymbolRequest>[];
    if (activeWidgets.contains(DashboardWidget.keyMarkets)) {
      for (final String symbol in marketTickerSettings.symbols) {
        try {
          requests.add(MarketRefreshTrackedSymbolRequest(symbol: normalizeSymbol(symbol), kind: MarketRefreshTrackedSymbolKind.market));
        } catch (_) {}
      }
    }
    if (activeWidgets.contains(DashboardWidget.currencies)) {
      for (final String symbol in currencySettings.symbols) {
        try {
          requests.add(MarketRefreshTrackedSymbolRequest(symbol: normalizeCurrencyCode(symbol), kind: MarketRefreshTrackedSymbolKind.currency));
        } catch (_) {}
      }
    }
    for (final Holding holding in holdings) {
      if (holding.isHidden) {
        continue;
      }
      try {
        requests.add(MarketRefreshTrackedSymbolRequest(symbol: normalizeSymbol(holding.symbol), kind: MarketRefreshTrackedSymbolKind.market));
      } catch (_) {}
    }
    final Set<String> seen = <String>{};
    return requests.where((MarketRefreshTrackedSymbolRequest request) => seen.add(request.symbol)).toList();
  }
}

File: swift/Sources/MoneyHero/MarketData/MarketRefreshRequestPlanner.swift
enum MarketRefreshTrackedSymbolKind: Sendable {
    case market
    case currency
}

struct MarketRefreshTrackedSymbolRequest: Sendable {
    let symbol: String
    let kind: MarketRefreshTrackedSymbolKind
}

enum MarketRefreshRequestPlanner {
    static func trackedSymbols(
        activeWidgets: [DashboardWidget],
        holdings: [Holding],
        currencySettings: CurrencyWidgetSetting,
        marketTickerSettings: MarketTickerSetting
    ) -> [String] {
        trackedRequests(
            activeWidgets: activeWidgets,
            holdings: holdings,
            currencySettings: currencySettings,
            marketTickerSettings: marketTickerSettings
        )
        .map(\.symbol)
    }

    static func trackedRequests(
        activeWidgets: [DashboardWidget],
        holdings: [Holding],
        currencySettings: CurrencyWidgetSetting,
        marketTickerSettings: MarketTickerSetting
    ) -> [MarketRefreshTrackedSymbolRequest] {
        var requests: [MarketRefreshTrackedSymbolRequest] = []

        if activeWidgets.contains(.keyMarkets) {
            requests.append(
                contentsOf: marketTickerSettings.symbols.compactMap { symbol in
                    guard let normalized = try? normalizeSymbol(symbol) else {
                        return nil
                    }
                    return MarketRefreshTrackedSymbolRequest(symbol: normalized, kind: .market)
                }
            )
        }

        if activeWidgets.contains(.currencies) {
            requests.append(
                contentsOf: currencySettings.symbols.compactMap { symbol in
                    guard let normalized = normalizedCurrencySymbol(symbol) else {
                        return nil
                    }
                    return MarketRefreshTrackedSymbolRequest(symbol: normalized, kind: .currency)
                }
            )
        }

        for holding in holdings {
            guard !holding.isHidden else {
                continue
            }
            if let normalized = try? normalizeSymbol(holding.symbol) {
                requests.append(MarketRefreshTrackedSymbolRequest(symbol: normalized, kind: .market))
            }
        }

        var seen: Set<String> = []
        return requests.filter { seen.insert($0.symbol).inserted }
    }

    private static func normalizedCurrencySymbol(_ symbol: String) -> String? {
        try? normalizeCurrencyCode(symbol)
    }
}
